rm/rm-client/src/main/java/org/ow2/proactive/resourcemanager/common/event/RMInitialState.java

```diff
@@ -61,99 +57,87 @@
+        toAdd.forEach(rmEvent -> events.add(rmEvent));
+        latestCounter.set(Math.max(latestCounter.get(), findLargestCounter(events.getSortedItems())));
     }
 
     /**
      * Current version of RM portal and maybe other clients expects "nodesEvents" inside JSON
      *
-     * @return
+     * @return list of RMNodeEvent
      */
     public List<RMNodeEvent> getNodesEvents() {
-        return Collections.unmodifiableList(new ArrayList<>(this.nodeEvents.values()));
+        return getNodesEvents(events.getSortedItems());
     }
 
     /**
      * Current version of RM portal and maybe other clients expects "nodeSource" inside JSON
      *
-     * @return
+     * @return list of RMNodeSourceEvent
      */
     public List<RMNodeSourceEvent> getNodeSource() {
-        return Collections.unmodifiableList(new ArrayList<>(this.nodeSourceEvents.values()));
+        return getNodeSource(events.getSortedItems());
+    }
+
+    public List<RMNodeEvent> getNodesEvents(Collection<RMEvent> rmEvents) {
+        return rmEvents.stream()
+                       .filter(event -> event instanceof RMNodeEvent)
+                       .map(event -> (RMNodeEvent) event)
+                       .collect(Collectors.toList());
+    }
+
+    public List<RMNodeSourceEvent> getNodeSource(Collection<RMEvent> rmevents) {
+        return rmevents.stream()
+                       .filter(event -> event instanceof RMNodeSourceEvent)
+                       .map(event -> (RMNodeSourceEvent) event)
+                       .collect(Collectors.toList());
     }
 
     public long getLatestCounter() {
         return latestCounter.get();
     }
 
     public void nodeAdded(RMNodeEvent event) {
-        updateCounter(event);
-        nodeEvents.put(event.getNodeUrl(), event);
-
+        update(event);
     }
 
     public void nodeStateChanged(RMNodeEvent event) {
-        updateCounter(event);
-        nodeEvents.put(event.getNodeUrl(), event);
+        update(event);
     }
 
     public void nodeRemoved(RMNodeEvent event) {
-        updateCounter(event);
-        nodeEvents.put(event.getNodeUrl(), event);
+        update(event);
     }
 
     public void nodeSourceAdded(RMNodeSourceEvent event) {
-        updateCounter(event);
-        nodeSourceEvents.put(event.getSourceName(), event);
+        update(event);
     }
 
     public void nodeSourceRemoved(RMNodeSourceEvent event) {
-        updateCounter(event);
-        nodeSourceEvents.put(event.getSourceName(), event);
+        update(event);
     }
 
     public void nodeSourceStateChanged(RMNodeSourceEvent event) {
-        updateCounter(event);
-        nodeSourceEvents.put(event.getSourceName(), event);
+        update(event);
     }
 
-    private void updateCounter(RMEvent event) {
-        latestCounter.set(Math.max(latestCounter.get(), event.getCounter()));
+    protected void update(RMEvent event) {
+        events.add(event);
+        updateCounter(event);
```

Comment:
Is there a reason why before we would update the counter before adding the event, and now, we do the opposite?

Reply:
No, there is no reason, it is just like this